osquery/tables/system/windows/windows_eventlog.cpp

```diff
@@ -0,0 +1,219 @@
+  row["channel"] = SQL_TEXT(windows_event.source);
+  row["provider_name"] = SQL_TEXT(windows_event.provider_name);
+  row["provider_guid"] = SQL_TEXT(windows_event.provider_guid);
+  row["eventid"] = INTEGER(windows_event.event_id);
+  row["task"] = INTEGER(windows_event.task_id);
+  row["level"] = INTEGER(windows_event.level);
+  row["pid"] = INTEGER(windows_event.pid);
+  row["tid"] = INTEGER(windows_event.tid);
+
+  row["keywords"] = SQL_TEXT(windows_event.keywords);
+  row["data"] = SQL_TEXT(windows_event.data);
+
+  if (context.hasConstraint("time_range", EQUALS)) {
+    auto time_range = context.constraints["time_range"].getAll(EQUALS);
+    row["time_range"] = SQL_TEXT(*time_range.begin());
+  } else {
+    row["time_range"] = SQL_TEXT("");
+  }
+
+  if (context.hasConstraint("timestamp", EQUALS)) {
+    auto timestamp = context.constraints["timestamp"].getAll(EQUALS);
+    row["timestamp"] = SQL_TEXT(*timestamp.begin());
+  }
+
+  results.push_back(row);
+}
+
+void parseQueryResults(QueryContext& context,
+                       EVT_HANDLE queryResults,
+                       QueryData& results) {
+  std::vector<EVT_HANDLE> events(kNumEventsBlock);
+  unsigned long numEvents = 0;
+
+  // Retrieve the events one block at a time
+  auto ret = EvtNext(
+      queryResults, kNumEventsBlock, events.data(), INFINITE, 0, &numEvents);
+
+  while (ret != FALSE) {
+    for (unsigned long i = 0; i < numEvents; i++) {
+      std::vector<wchar_t> renderedContent;
+      unsigned long renderedBuffSize = 0;
+      unsigned long renderedBuffUsed = 0;
+      unsigned long propCount = 0;
+      EvtRender(nullptr,
+                events[i],
+                EvtRenderEventXml,
+                renderedBuffSize,
+                renderedContent.data(),
+                &renderedBuffUsed,
+                &propCount);
+
+      if (GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+        renderedBuffSize = renderedBuffUsed;
+        renderedContent.resize(renderedBuffSize);
+        EvtRender(nullptr,
+                  events[i],
+                  EvtRenderEventXml,
+                  renderedBuffSize,
+                  renderedContent.data(),
+                  &renderedBuffUsed,
+                  &propCount);
+      }
+      if (GetLastError() != ERROR_SUCCESS) {
+        LOG(WARNING) << "Failed to render windows event with "
+                     << GetLastError();
+        continue;
+      }
+
+      std::wstringstream xml_event;
+      xml_event << renderedContent.data();
+      parseWelXml(context, xml_event.str(), results);
+      EvtClose(events[i]);
+    }
+
+    ret = EvtNext(
+        queryResults, kNumEventsBlock, events.data(), INFINITE, 0, &numEvents);
+  }
+}
+
+void genXfilterFromConstraints(QueryContext& context, std::string& xfilter) {
+  std::vector<std::string> xfilterList;
+
+  auto eids = context.constraints["eventid"].getAll(EQUALS);
+  if (!eids.empty()) {
+    xfilterList.emplace_back(
```

Comment:
Is it OK that a query can "inject" into this filter list? What is the worst thing that could happen from this interpolation?

Reply:
The number of elements in `xfilterList` will be limited by the constraints. One can make the element string too large by providing too many constraints with the query but that will be limited by query string size. 
Could you let me know your thoughts about it? I might have missed something.

Reply:
This filter is passed to [EvtQuery](https://docs.microsoft.com/en-us/windows/win32/api/winevt/nf-winevt-evtquery), which seems to be read-only. Does that alleviate your concerns @theopolis?

Reply:
I just pinged @theopolis on Slack to see if we can wrap this up.